src/backend/binaryen/utils.ts

```diff
@@ -1587,123 +1673,104 @@ export namespace FunctionalFuncs {
-    META_INDEX = 0,
-}
-
-export const SIZE_OF_META_FIELD = 12;
-
-export const enum MetaDataOffset {
-    TYPE_ID_OFFSET = 0,
-    IMPL_ID_OFFSET = 4,
-    COUNT_OFFSET = 8,
-    FIELDS_PTR_OFFSET = 12,
-}
+    export function getWASMObjectVtable(
+        module: binaryen.Module,
+        ref: binaryen.ExpressionRef,
+    ) {
+        return getWasmStructFieldByIndex(
+            module,
+            ref,
+            baseVtableType.typeRef,
+            StructFieldIndex.VTABLE_INDEX,
+        );
+    }
 
-export const enum MetaFieldOffset {
-    NAME_OFFSET = 0,
-    FLAG_AND_INDEX_OFFSET = 4,
-    TYPE_OFFSET = 8,
-}
+    export function getWASMObjectMeta(
+        module: binaryen.Module,
+        ref: binaryen.ExpressionRef,
+    ) {
+        const vtable = getWASMObjectVtable(module, ref);
+        return getWasmStructFieldByIndex(
+            module,
+            vtable,
+            binaryen.i32,
+            VtableFieldIndex.META_INDEX,
+        );
+    }
 
-export function getFieldFromMetaByOffset(
-    module: binaryen.Module,
-    meta: binaryen.ExpressionRef,
-    offset: number,
-) {
-    return module.i32.load(offset, memoryAlignment, meta);
-}
+    export function isUndefinedIndex(
+        module: binaryen.Module,
+        indexRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(indexRef, module.i32.const(268435455));
+    }
 
-export function getWasmStructFieldByIndex(
-    module: binaryen.Module,
-    ref: binaryen.ExpressionRef,
-    typeRef: binaryen.Type,
-    idx: number,
-) {
-    return binaryenCAPI._BinaryenStructGet(
-        module.ptr,
-        idx,
-        ref,
-        typeRef,
-        false,
-    );
-}
+    export function isUndefinedFlag(
+        module: binaryen.Module,
+        flagRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(flagRef, module.i32.const(15));
+    }
 
-export function getWASMObjectVtable(
-    module: binaryen.Module,
-    ref: binaryen.ExpressionRef,
-) {
-    return getWasmStructFieldByIndex(
-        module,
-        ref,
-        baseVtableType.typeRef,
-        StructFieldIndex.VTABLE_INDEX,
-    );
-}
+    export function isFieldFlag(
+        module: binaryen.Module,
+        flagRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(flagRef, module.i32.const(ItableFlag.FIELD));
+    }
 
-export function getWASMObjectMeta(
-    module: binaryen.Module,
-    ref: binaryen.ExpressionRef,
-) {
-    const vtable = getWASMObjectVtable(module, ref);
-    return getWasmStructFieldByIndex(
-        module,
-        vtable,
-        binaryen.i32,
-        VtableFieldIndex.META_INDEX,
-    );
-}
+    export function isMethodFlag(
+        module: binaryen.Module,
+        flagRef: binaryen.ExpressionRef,
+    ) {
+        return module.i32.eq(flagRef, module.i32.const(ItableFlag.METHOD));
+    }
 
-export interface SourceMapLoc {
-    location: SourceLocation;
-    ref: binaryen.ExpressionRef;
+    export function isTypeIdEqual(
```

Comment:
You check both typeid and implementid, so the function name `isTypeIdEqual` is not reasonable.
Had better rename it to `isShapeCompatible`

Reply:
Get it.